documentation/asciidoc/computers/raspberry-pi/raspberry-pi-industrial.adoc

```diff
@@ -92,7 +92,63 @@ It is possible to prevent the customer OTP bits from being read at all. This can
 
 [,bash]
 ----
-/opt/vc/bin/vcmailbox 0x00038021 8 8 0xffffffff 0xaffebabe
+vcmailbox 0x00038021 8 8 0xffffffff 0xaffebabe
 ----
 
 This operation is unlikely to be useful for the vast majority of users, and is irreversible.
+
+=== Device specific private key
+Eight rows of OTP (256bits) are available for use as a device-specific private key. This is intended to support file-system encryption.
+
+These rows can be programmed and read using similar `vcmailbox` commands to those used for managing customer OTP rows. If
+secure-boot / file-system encryption is not required then the device private key rows can be used to store general purpose information.
+
+* The device private key rows can only be read via the `vcmailbox` command which requires access to `/dev/vcio` which is restricted to the `video` group on Raspbvrry Pi OS.
+* Raspberry Pi 4 and earlier do not have a hardware protected key store.  It is recommended that these feature is used in conjunction with https://github.com/raspberrypi/usbboot/blob/master/secure-boot-example/README.md[secure-boot] in order to restrict access to this data.
+* Raspberry Pi OS does not support an encrypted root-filesystem.
+
+See https://gitlab.com/cryptsetup/cryptsetup[cryptsetup] for more information about open-source disk encryption.
+
+==== Key programming script `rpi-otp-private-key`
+The https://github.com/raspberrypi/usbboot/blob/master/tools/rpi-otp-private-key[rpi-otp-private-key] script wraps the device private key `vcmailbox` APIs in order to make it easier to read/write a key in the same format as OpenSSL.
+
+Read the key as a 64-byte hex number
+[,bash]
+----
+rpi-otp-private-key
+----
+
+Example output
+----
+f8dbc7b0a4fcfb1d706e298ac9d0485c2226ce8df7f7596ac77337bd09fbe160
+----
+
+Writes a 64-byte randomly generated number to the device private key. +
+**Warning: This operation cannot be undone. **
+[,bash]
+----
+# rpi-otp-private-key -w $(openssl rand -hex 32)
+----
+
+==== Mailbox API for reading/writing the key.
+Read all of the rows.
+[,bash]
+----
+vcmailbox vcmailbox 0x00030081 40 40 0 8 0 0 0 0 0 0 0 0
+----
+
+Example output
+----
+0x00000040 0x80000000 0x00030081 0x00000028 0x80000028 0x00000000 0x00000008 0xf8dbc7b0 0xa4fcfb1d 0x706e298a 0xc9d0485c 0x2226ce8d 0xf7f7596a 0xc77337bd 0x09fbe160 0x00000000
+----
+
+Write all of the row (replace the trailing eight zeros with the key data)
+[,bash]
+----
+vcmailbox vcmailbox 0x00038081 40 40 0 8 0 0 0 0 0 0 0 0
```

Comment:
This seems to be identical to the command above where you read all the rows? :confused:

Reply:
No, this is using the standard set parameter bit so 0x0003**8**081  instead of 0x00030081

Reply:
Ahhh, sneaky :laughing:

Reply:
It's more obvious in the source code ;) But also one of the reasons for writing a wrapper script